### ISSUE TITLE
check if network data consumption can be optimized

### ISSUE THREAD

**Issue by adbenitez** [deltachat/deltachat-core#572]
I was performing some tests, with DeltaChat 0.100 and K9 Mail, sending a simple(and unencrypted)  message "Hello world!!!" with DeltaChat uses ~12KB  while sending the same message using K9 Mail uses only 2KB of network flow(so I can send 6 messages with K9 for every email I send with DC??). This seems to be related to the fact that DC sends a copy of the email to the user itself and have to move it to the DeltaChat folder, is this message downloaded??? if so, this is not much better than the previous approach.

- Delta Chat Version: 0.100

**Comment by hpk42** [comment]
On Sun, Feb 10, 2019 at 08:42 -0800, Asiel Díaz Benítez wrote:
> I was performing some tests, with DeltaChat 0.100 and K9 Mail, sending a simple(and unencrypted)  message "Hello world!!!" with DeltaChat uses ~12KB  while sending the same message using K9 Mail uses only 2KB of network flow(so I can send 6 messages with K9 for every email I send with DC??). This seems to be related to the fact that DC sends a copy of the email to the user itself and have to move it to the DeltaChat folder, is this message downloaded??? if so, this is not much better than the previous approach.

no, this should not happen.  the two message sizes should be more similar. 
Are both one:one mails?  could you mail me with the two progrs at my chat address
so i can verify what's going on?

**Comment by adbenitez** [comment]
yes, they were both one:one emails, I think main problem is headers overhead, plus the moving stuff.

**Comment by testbird** [comment]
> [always sending key is] not a waste. it's to allow any random, also new client, to encrypt to you;

That may be ok for classic email. For quick imap-push chat messaging it may be ok though, not to send the keys every time, allow for new clients to request sending a key, and answering the request with the key if it is from a known contact.

**Comment by adbenitez** [comment]
also if I have encryption on but the other peer have it off, there are no need to send the key

**Comment by testbird** [comment]
Hm yes, I thought the peer may not be able to switch it on again anytime, and currently only always sending the key allows this. But...

The idea in the last post was to allow to switch it on anytime, by having a way to request the current key.

However, requesting keys has drawbacks:
* requires (timely) network connectivity
* requires background system messages (preferable to avoid this, also unnecessary once proper key exchange is possible with out-of-band QR-codes, etc.)

And as IMAP is designed to sync to newly installed MUA clients to the current state, requesting keys is completely unnecessary.

Thus, the trick is simply to store and sync the known contacts with their keys using email (files) in the chat subfolder on the imap server. The coi design https://github.com/coi-dev/coi-specs/blob/master/coi-client-spec.md#contact-storage-messages contains some good ideas.

Well, it's still time to reconsider the MPL relicencing, as it provides no immunity against propriatary leachers and propriatary feature extension forks (product leveraging).
